util/pkg/vfs/s3context.go

```diff
@@ -106,26 +111,29 @@ func getCustomS3Config(endpoint string, region string) (*aws.Config, error) {
 	return s3Config, nil
 }
 
-func (s *S3Context) getRegionForBucket(bucket string) (string, error) {
-	region := func() string {
+func (s *S3Context) getDetailsForBucket(bucket string) (*S3BucketDetails, error) {
+	bucketDetails := func() *S3BucketDetails {
 		s.mutex.Lock()
 		defer s.mutex.Unlock()
-		return s.bucketLocations[bucket]
+		return s.bucketDetails[bucket]
 	}()
 
-	if region != "" {
-		return region, nil
+	if bucketDetails != nil && bucketDetails.region != "" {
+		return bucketDetails, nil
 	}
 
+	bucketDetails = &S3BucketDetails{region: "", defaultEncryption: false}
+
 	// Probe to find correct region for bucket
 	endpoint := os.Getenv("S3_ENDPOINT")
 	if endpoint != "" {
 		// If customized S3 storage is set, return user-defined region
-		region = os.Getenv("S3_REGION")
-		if region == "" {
-			region = "us-east-1"
+		bucketDetails.region = os.Getenv("S3_REGION")
+		if bucketDetails.region == "" {
+			bucketDetails.region = "us-east-1"
+			bucketDetails.defaultEncryption = s.checkDefaultEncryption(bucketDetails.region, bucket)
 		}
-		return region, nil
+		return bucketDetails, nil
```

Comment:
I don't think we set defaultEncryption if S3_REGION is set?  Also we aren't caching it here...